apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -329,7 +355,17 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
    * @override
    */
   protected serialize(): string {
-    return yaml.safeDump(this._shrinkwrapJson, SHRINKWRAP_YAML_FORMAT);
+    let shrinkwrapContent: string = yaml.safeDump(this._shrinkwrapJson, SHRINKWRAP_YAML_FORMAT);
+    if (this._shrinkwrapHash) {
+      // We need to validate that the hashes are aligned before writing.
+      if (this._shrinkwrapHash !== crypto.createHash('sha1').update(shrinkwrapContent).digest('hex')) {
+        throw new Error('Content hash does not match');
```

Comment:
It seems a little odd that `PnpmShrinkwrapFile.serialize()` is throwing an error about a mismatched hash.  The job of `serialize()` is to write out the file, so intuitively I would expect it to generate a correct hash as part of the operation.

Instead, I would expect the validity of the hash to be checked during _loading_, so we could have an operation like:

```ts
  if (!shrinkwrapFile.hashIsValid) {
    throw new Error('Manual modifications were detected in the shrinkwrap file.  Please run "rush update".');
  }
```

or perhaps:

```ts
  public validateHash(): void {
    if (!this._hashIsValid) {
      throw new Error('Manual modifications were detected in the shrinkwrap file.  Please run "rush update".');
    }
  }
```

Reply:
This works as well. Only reason I kept it like this was to ensure that the hash being written was correct. I guess my question would be then, do we write the hash of the content, or the value of the field? Is it preferred that the getter for the hash calculates the hash on access so that it's always correct?

Reply:
I don't see any reason to support serializing an incorrect hash.

Either `this._shrinkwrapHash` can get recalculated if `serialize()` is called, or `this._shrinkwrapHash` can reflect the value loaded from disk (ignoring the value calculated during `serialize()` ).  Either way is fine, as long as it's documented in a code comment.